### PR TITLE
PR adoption terminal idempotency hardening

### PR DESCRIPTION
**Pull request by dimileeh**
Automatically opened by AWF workspace `ws_e5b86a598da842e0aaf50d1f` (agent: `codex`, model: `gpt-5.5`, effort: `xhigh`).

### Task
Implement the P1 backlog slice "Harden PR monitor adoption when a previous repo/PR adoption workspace is terminal" from TODO/pre-gke-industrial-readiness.md.

Context and incident:
- An operator tried to adopt PR #206 after its previous AWF adoption workspace had been destroyed.
- AWF used the deterministic repo/PR adoption idempotency key and returned the old destroyed workspace as a successful attached_existing response.
- remonitor correctly rejected that destroyed workspace, so the operator had to manually clear the stale idempotency claim to create a new monitor.

Required behavior:
- Adoption must attach only to live adoption workspaces that can still monitor the PR.
- Terminal adoption rows such as destroyed, cancelled, failed, and superseded must not satisfy the deterministic repo/PR idempotency key as a successful monitor attachment.
- For an open PR with only terminal adoption history, AWF must create a fresh monitor workspace while preserving audit/lineage context and avoiding unique idempotency conflicts.
- Active monitor policy mismatches must still return PR_ADOPTION_POLICY_CONFLICT.
- Concurrent adoption attempts must create at most one live monitor.
- REST, CLI, and MCP-visible behavior must remain consistent and structured.

Engineering requirements:
- Use strict TDD: write failing regression tests first, then implement the smallest correct fix.
- No fake tests, empty assertions, monkeypatches that skip the behavior under test, or lowered validation/coverage/profile gates.
- Add focused unit coverage for destroyed, cancelled, failed, and superseded prior adoption rows; active idempotent reattach; active policy conflict; and concurrent adoption race behavior.
- Preserve 99%+ coverage and run focused validation first, then the broader Python/control-plane validation required by the touched surface.
- Keep changes scoped to the adoption service/API/CLI/MCP contract as needed; do not rewrite unrelated backlog or active ledger rows.
- Update TODO/pre-gke-industrial-readiness.md only with implementation-backed evidence for this slice.
- Commit changes and let AWF create and monitor the PR with auto_merge=true.

---
Validation: 5 profile command(s) passed inside the workspace container.

## Summary by CodeRabbit

* **New Features**
  * Enhanced PR monitor adoption to intelligently reattach to existing live monitoring workspaces when available, improving idempotency and adoption tracking.
  * Implemented fresh workspace creation when only historical adoption records exist, eliminating conflicts.

* **Bug Fixes**
  * Added validation for conflicting auto-merge policies with clear error responses (HTTP 409) when adoption policies mismatch.

* **Tests**
  * Expanded test coverage for adoption replay scenarios, terminal states, and concurrent adoption edge cases.

### CODE REVIEW

File: src/awf/service/pr_monitor_adoption.py [hunk 1]
```diff
@@ -102,21 +113,39 @@ async def adopt(
...
+        )
         workspace = await self._create_adoption_workspace(
             request=request,
             repo=repo,
             metadata=metadata,
-            idempotency_key=idempotency_key,
+            idempotency_key=workspace_idempotency_key,
+            logical_idempotency_key=idempotency_key,
+            previous_terminal_adoptions=previous_terminal_adoptions,
         )
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/service/pr_monitor_adoption.py:184]: fixed in commit c30ca94f1b96e419219af694ad3795c902606718

File: src/awf/db/repositories.py [hunk 2]
```diff
@@ -2450,6 +2450,44 @@ async def get_by_idempotency_key(self, key: str) -> Workspace | None:
...
+                    Workspace.task_external_id == task_external_id,
+                    Workspace.idempotency_key == idempotency_key,
+                    and_(
+                        Workspace.task_kind == task_kind,
+                        Workspace.pr_number == pr_number,
+                    ),
+                )
+            )
+            .order_by(Workspace.created_at.asc(), Workspace.id.asc())
+        )
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/db/repositories.py:2719]: fixed in commit 6a77b66e4262a88142cd3bf0b5a7eebf6f784d3e

File: src/awf/service/pr_monitor_adoption.py [hunk 3]
```diff
@@ -354,6 +395,68 @@ async def _default_metadata_fetcher(
...
+    *,
+    logical_idempotency_key: str,
+) -> str:
+    if await workspace_repo.get_by_idempotency_key(logical_idempotency_key) is None:
+        return logical_idempotency_key
+
+    for generation in range(1, 1000):
+        candidate = f"{logical_idempotency_key}:g{generation}"
+        if await workspace_repo.get_by_idempotency_key(candidate) is None:
+            return candidate
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/service/pr_monitor_adoption.py:585]: fixed in commit 24a1d308efdb2ca5f6b68a782f210df60ec57fdb

File: src/awf/service/pr_monitor_adoption.py [hunk 4]
```diff
@@ -354,6 +395,68 @@ async def _default_metadata_fetcher(
...
+        attempt = await attempt_repo.get_by_workspace_id(workspace.id)
+        lineage.append(
+            {
+                "workspace_id": workspace.id,
+                "status": workspace.status,
+                "task_id": attempt.task_id if attempt is not None else None,
+                "attempt_id": attempt.id if attempt is not None else None,
+            }
+        )
+    return lineage
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/service/pr_monitor_adoption.py]: fixed in commit 6a77b66e4262a88142cd3bf0b5a7eebf6f784d3e

File: src/awf/db/repositories.py [hunk 5]
```diff
@@ -2450,6 +2450,44 @@ async def get_by_idempotency_key(self, key: str) -> Workspace | None:
...
+                or_(
+                    Workspace.task_external_id == task_external_id,
+                    Workspace.idempotency_key == idempotency_key,
+                    and_(
+                        Workspace.task_kind == task_kind,
+                        Workspace.pr_number == pr_number,
+                    ),
+                )
+            )
+            .order_by(Workspace.created_at.asc(), Workspace.id.asc())
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/db/repositories.py:2718]: fixed in commit 7823a2b70ebd09db508741010525b576901000e6

File: src/awf/db/repositories.py [hunk 6]
```diff
@@ -3175,6 +3213,45 @@ async def add_events(
...
+    pr_number: int,
+) -> bool:
+    if workspace.task_kind == task_kind and workspace.task_external_id == task_external_id:
+        return True
+
+    adoption = _workspace_pr_adoption_policy(workspace)
+    if not adoption:
+        return False
+    if workspace.task_kind != task_kind and workspace.idempotency_key != idempotency_key:
+        return False
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/db/repositories.py:3475]: FALSE POSITIVE: `workspaces.task_external_id` is not unique and the first query branch can return a same-external-id row with a different `task_kind`, which this guard filters because the earlier return also requires the expected task kind.

File: src/awf/service/pr_monitor_adoption.py [hunk 7]
```diff
@@ -42,6 +44,17 @@
...
+_LIVE_ADOPTION_STATUSES = frozenset(
+    {
+        WorkspaceStatus.requested.value,
+        WorkspaceStatus.provisioning.value,
+        WorkspaceStatus.ready.value,
+        WorkspaceStatus.running.value,
+        WorkspaceStatus.validating.value,
+        WorkspaceStatus.pushing.value,
+        WorkspaceStatus.monitoring_pr.value,
+    }
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/service/pr_monitor_adoption.py:60]: fixed in commit 83f609fed64ee4978e18735f2f01a9da698961a5

File: src/awf/service/pr_monitor_adoption.py [hunk 8]
```diff
@@ -42,6 +44,18 @@
...
+        WorkspaceStatus.requested.value,
+        WorkspaceStatus.provisioning.value,
+        WorkspaceStatus.ready.value,
+        WorkspaceStatus.running.value,
+        WorkspaceStatus.validating.value,
+        WorkspaceStatus.pushing.value,
+        WorkspaceStatus.monitoring_pr.value,
+        WorkspaceStatus.destroying.value,
+    }
+)
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/service/pr_monitor_adoption.py:61]: fixed in commit 2c6da6c45965cf612ae1304b17a9be8b613ec9f5

File: src/awf/service/pr_monitor_adoption.py [hunk 9]
```diff
@@ -204,16 +241,32 @@ async def _create_adoption_workspace(
...
+        except TaskExternalIdConflictError:
+            if not previous_terminal_adoptions:
+                raise
+            workspace.task_external_id = _adoption_generation_external_id(
+                repo_slug=repo.slug(),
+                pr_number=metadata.number,
+                logical_idempotency_key=logical_idempotency_key,
+                workspace_idempotency_key=idempotency_key,
+            )
+            task = await _create_or_get_task(external_id=workspace.task_external_id)
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/service/pr_monitor_adoption.py]: fixed in commit 93d7a05be25c7d244ba71040ec13e70616d6f387

File: src/awf/service/pr_monitor_adoption.py [hunk 10]
```diff
@@ -354,6 +426,83 @@ async def _default_metadata_fetcher(
...
+
+
+async def _next_adoption_workspace_idempotency_key(
+    workspace_repo: WorkspaceRepository,
+    *,
+    logical_idempotency_key: str,
+) -> str:
+    existing_keys = set(await workspace_repo.list_idempotency_key_family(logical_idempotency_key))
+    if logical_idempotency_key not in existing_keys:
+        return logical_idempotency_key
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/service/pr_monitor_adoption.py]: fixed in commit ae5a74cd2e6107ba3d80558742a076bb880b6dae

File: src/awf/service/pr_monitor_adoption.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/service/pr_monitor_adoption.py:60]: fixed in commit 7543c69fc09d3a00f1ee0bc4548f676949446159

File: src/awf/service/pr_monitor_adoption.py [hunk 11]
```diff
@@ -102,21 +116,45 @@ async def adopt(
...
         metadata = await self._fetch_metadata(repo=repo, pr_number=pr_number)
+        task_idempotency_keys = await _task_idempotency_key_family(
+            self._session,
+            logical_idempotency_key=idempotency_key,
+        )
+        workspace_idempotency_key = await _next_adoption_workspace_idempotency_key(
+            workspace_repo,
+            logical_idempotency_key=idempotency_key,
+            reserved_idempotency_keys=task_idempotency_keys,
+            require_generation=bool(previous_terminal_adoptions),
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/service/pr_monitor_adoption.py]: fixed in commit 35accd7bf7c2055b2a7f4a7b98036112d70d00ed

**Comment by dimileeh** [comment]
@coderabbitai review

File: docs/awf-plans/ws_e5b86a598da842e0aaf50d1f.md [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+
+6. MCP: visible behavior matches REST/service
+   - Through `awf_adopt_pull_request_monitor`, assert terminal prior adoption returns a normal structured success payload for the fresh workspace.
+   - Assert active policy conflict returns `CallToolResult.isError is True` with structured `PR_ADOPTION_POLICY_CONFLICT`.
+
+## Implementation Approach
+
+1. Define adoption liveness vocabulary in `src/awf/service/pr_monitor_adoption.py`.
+   - Live adoption statuses should be the statuses that can still progress into or run PR monitoring: `requested`, `provisioning`, `ready`, `running`, `validating`, `pushing`, and `monitoring_pr`.
+   - Terminal/non-attach statuses should include at least `completed`, `failed`, `cancelled`, `destroying`, `destroyed`, and raw string `superseded`. The tests required by this slice will cover `destroyed`, `cancelled`, `failed`, and `superseded`.
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply docs/awf-plans/ws_e5b86a598da842e0aaf50d1f.md]: fixed in commit 99d4493d7a4f7747734c7a042ef4297634b48b3e

File: src/awf/service/pr_monitor_adoption.py [hunk 13]
```diff
@@ -102,21 +116,55 @@ async def adopt(
...
+        for workspace in adoption_history:
+            if workspace.idempotency_key is not None:
+                workspace_idempotency_keys.append(workspace.idempotency_key)
+        workspace_idempotency_key = await _next_adoption_workspace_idempotency_key(
+            workspace_repo,
+            logical_idempotency_key=idempotency_key,
+            known_workspace_keys=workspace_idempotency_keys,
+            reserved_idempotency_keys=[*task_idempotency_keys, *task_external_id_keys],
+            require_generation=bool(previous_terminal_adoptions),
+        )
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/service/pr_monitor_adoption.py]: fixed in commit e0c566ed

File: tests/unit/service/test_pr_monitor_adoption.py [hunk 14]
```diff
@@ -85,6 +99,24 @@ async def _count(session: AsyncSession, model: type[Any]) -> int:
...
+    external_id = adoption_module._adoption_external_id(
+        repo_slug=repo_slug,
+        pr_number=pr_number,
+    )
+    rows = await session.execute(
+        select(Workspace)
+        .where(Workspace.task_external_id == external_id)
+        .order_by(Workspace.created_at.asc(), Workspace.id.asc())
+    )
+    return list(rows.scalars())
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply tests/unit/service/test_pr_monitor_adoption.py]: fixed in commit 8c3d18e9

File: docs/awf-plans/ws_e5b86a598da842e0aaf50d1f.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (dimileeh) [reply docs/awf-plans/ws_e5b86a598da842e0aaf50d1f.md]: fixed in commit b5dd106ed749fc8e60692f088337557a202121b3

[the remark this answers is not in the document]
Comment (dimileeh) [reply docs/awf-plans/ws_e5b86a598da842e0aaf50d1f.md]: fixed in commit b5dd106ed749fc8e60692f088337557a202121b3

[the remark this answers is not in the document]
Comment (dimileeh) [reply docs/awf-plans/ws_e5b86a598da842e0aaf50d1f.md]: fixed in commit b5dd106ed749fc8e60692f088337557a202121b3

[the remark this answers is not in the document]
Comment (dimileeh) [reply docs/awf-plans/ws_e5b86a598da842e0aaf50d1f.md]: fixed in commit b5dd106ed749fc8e60692f088337557a202121b3

File: src/awf/service/pr_monitor_adoption.py [hunk 15]
```diff
@@ -295,16 +329,50 @@ async def _create_adoption_workspace(
...
+                idempotency_key=task_idempotency_key,
+                task_class=workspace.task_class,
+                owned_paths=list(workspace.owned_paths),
+            )
+
+        try:
+            task = await _create_or_get_task(
+                external_id=workspace.task_external_id,
+                task_idempotency_key=idempotency_key,
+            )
```
[the remark this answers is not in the document]
Comment (dimileeh) [reply src/awf/service/pr_monitor_adoption.py]: Fixed in d1a8f69. The regression now reproduces the stale state where the previous terminal workspace key is cleared, the old task still owns the logical idempotency key, and its external id has already been superseded. PR adoption now detects when that logical-key task already has an attempt and allocates the next task generation instead of linking the fresh monitor to stale task scope/title.
